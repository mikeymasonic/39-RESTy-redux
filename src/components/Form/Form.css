.Form {
  margin: auto;
  width: 50rem;
  background-color: purple;
  color: whitesmoke;
  padding: 20px;

  input[type=text] {
    font-size: 15px;
    width: 100%;
    margin-bottom: 10px;
  }
}

h3 {
  text-align: center;
}

p {
  padding: 1rem 2rem;
  line-height: 1.5rem;
}

textarea {
  display: block;
  margin: auto;
  width: 100%;
  height: 100px;
}

input[type=radio] {
  margin-bottom: 10px;
}