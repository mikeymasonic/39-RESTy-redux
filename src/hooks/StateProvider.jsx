import React, { createContext, useReducer, useContext } from 'react';
import PropTypes from 'prop-types';

const StateContext = createContext();

const initialState = {
  url: '',
  method: 'GET',
  body: '',
  disable: true,
  headers: {},
  response: {},
  requests: []
};

export function reducer(state, action) {
  switch(action.type) {
    case 'SET_URL':
      return { ...state, url: action.payload };
    case 'SET_METHOD':
      return { ...state, method: action.payload };
    case 'SET_BODY':
      return { ...state, body: action.payload };
    case 'SET_REQUESTS':
      return { ...state, requests: action.payload };
    case 'SET_DISABLE':
      return { ...state, disable: action.payload };
    case 'SET_RESPONSE':
      return { ...state, response: action.payload };
    case 'SET_HEADERS':
      return { ...state, headers: action.payload };
    
    default:
      return state;
  }
}

export const StateProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StateContext.Provider value={{ state, dispatch }}>
      {children}
    </StateContext.Provider>
  );
};

StateProvider.propTypes = {
  children: PropTypes.node
};

export const useGlobalState = () => {
  const { state } = useContext(StateContext);
  return state;
};

export const useURL = () => {
  const { url } = useGlobalState();
  return url;
};

export const useBody = () => {
  const { body } = useGlobalState();
  return body;
};

export const useMethod = () => {
  const { method } = useGlobalState();
  return method;
};

export const useDispatch = () => {
  const { dispatch } = useContext(StateContext);
  return dispatch;
};

export const useDisable = () => {
  const { disable } = useGlobalState();
  return disable;
};

export const useResponse = () => {
  const { response } = useGlobalState();
  return response;
};

export const useHeaders = () => {
  const { headers } = useGlobalState();
  return headers;
};

export const useRequests = () => {
  const { requests } = useGlobalState();
  return requests;
};
