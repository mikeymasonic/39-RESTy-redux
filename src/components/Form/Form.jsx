import React from 'react';
import PropTypes from 'prop-types';
import styles from './Form.css';
import { useDisable, useURL, useMethod, useBody } from '../../hooks/StateProvider';
import { useDispatch } from 'react-redux';

const Form = ({ onChange, onSubmit }) => {
  const dispatch = useDispatch();
  const [url, setUrl] = useState('');
  const [method, setMethod] = useState('');
  const [body, setBody] = useState('');
  const [disable, setDisable] = useState(false);

  const url = useURL();
  const method = useMethod();
  const body = useBody();
  const disable = useDisable();

  return (
    <div className={styles.Form}>
      <form onSubmit={onSubmit}>
        <label>URL:<input type="text" name="url" value={url} onChange={onChange}></input></label> 
        <div>
          <label><input type="radio" name="method" value="GET" checked={method === 'GET'} onChange={onChange} />GET</label>
          <label><input type="radio" name="method" value="POST" checked={method === 'POST'} onChange={onChange} />POST</label>
          <label><input type="radio" name="method" value="PUT" checked={method === 'PUT'} onChange={onChange} />PUT</label>
          <label><input type="radio" name="method" value="PATCH" checked={method === 'PATCH'} onChange={onChange} />PATCH</label>
          <label><input type="radio" name="method" value="DELETE" checked={method === 'DELETE'} onChange={onChange} />DELETE</label>
        </div>
        <textarea placeholder="JSON Body" name="body" value={body} onChange={onChange} disabled={disable}></textarea>
        <button type="submit">Submit</button>
      </form>
    </div>
  );
};

Form.propTypes = {
  url: PropTypes.string,
  method: PropTypes.string,
  body: PropTypes.string,
  disable: PropTypes.bool,
  onChange: PropTypes.func,
  onSubmit: PropTypes.func
};

export default Form;
