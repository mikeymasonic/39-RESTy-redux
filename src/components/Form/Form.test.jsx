import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';
import { StateProvider } from '../../hooks/StateProvider';

describe('Form component', () => {
  it('renders Form', () => {
    const wrapper = shallow(
      <StateProvider>
        <Form />
      </StateProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
