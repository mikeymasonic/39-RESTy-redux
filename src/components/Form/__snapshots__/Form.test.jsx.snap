// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "body": "",
        "disable": true,
        "headers": Object {},
        "method": "GET",
        "requests": Array [],
        "response": Object {},
        "url": "",
      },
    }
  }
>
  <Form />
</ContextProvider>
`;
